### DISCUSSION TITLE
Error cross compiling the library with yocto toolchain

### DISCUSSION THREAD

**Discussion by martxel** [aws/aws-iot-device-sdk-cpp-v2#765] [Q&A] [answered]
Hello.

I'm having some errors when trying to build the library with a toolchain generated with yocto.

I can successfully build and use the library for x86_64 with the following commands:
```
mkdir /home/user/libraries/x86_64/Debug/aws-iot-device-sdk-cpp-v2
cd /home/user/src/
git clone --recursive https://github.com/aws/aws-iot-device-sdk-cpp-v2.git
cd aws-iot-device-sdk-cpp-v2
git checkout v1.34.0
mkdir build && cd build
cmake .. -DCMAKE_INSTALL_PREFIX="/home/user/libraries/x86_64/Debug/aws-iot-device-sdk-cpp-v2/" -DCMAKE_BUILD_TYPE="Debug" -DUSE_OPENSSL=ON
cmake --build . --target install
```

Then, I repeat the steps for aarch64, using the toolchain created by yocto:
```
mkdir /home/user/libraries/aarch64/Debug/aws-iot-device-sdk-cpp-v2/
cd /home/user/src/aws-iot-device-sdk-cpp-v2
rm -fr build && mkdir build && cd build
source /opt/tdx-xwayland/6.8.0/environment-setup-cortexa53-tdx-linux
cmake .. -DCMAKE_INSTALL_PREFIX="/home/user/libraries/aarch64/Debug/aws-iot-device-sdk-cpp-v2/" -DCMAKE_BUILD_TYPE="Debug" -DUSE_OPENSSL=ON -DCMAKE_TOOLCHAIN_FILE="/opt/tdx-xwayland/6.8.0/sysroots/x86_64-tdxsdk-linux/usr/share/cmake/OEToolchainConfig.cmake"
cmake --build . --target install
```
When running the folllowing command (output in attached output.log):
```
cmake .. -DCMAKE_INSTALL_PREFIX="/home/user/libraries/aarch64/Debug/aws-iot-device-sdk-cpp-v2/" -DCMAKE_BUILD_TYPE="Debug" -DUSE_OPENSSL=ON -DCMAKE_TOOLCHAIN_FILE="/opt/tdx-xwayland/6.8.0/sysroots/x86_64-tdxsdk-linux/usr/share/cmake/OEToolchainConfig.cmake"
```
It builds and installs S2N correctly.

However, when running the build command, it builds some libraries correctly until I get this error:
```
[ 33%] Building C object crt/aws-crt-cpp/crt/aws-c-io/CMakeFiles/aws-c-io.dir/source/s2n/s2n_tls_channel_handler.c.o
/home/user/src/aws-iot-device-sdk-cpp-v2/crt/aws-crt-cpp/crt/aws-c-io/source/s2n/s2n_tls_channel_handler.c:23:10: fatal error: s2n.h: No such file or directory
   23 | #include <s2n.h>
      |          ^~~~~~~
compilation terminated.
gmake[2]: *** [crt/aws-crt-cpp/crt/aws-c-io/CMakeFiles/aws-c-io.dir/build.make:342: crt/aws-crt-cpp/crt/aws-c-io/CMakeFiles/aws-c-io.dir/source/s2n/s2n_tls_channel_handler.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1354: crt/aws-crt-cpp/crt/aws-c-io/CMakeFiles/aws-c-io.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

It looks like it can't find the s2n.h header.

The full output with the error is in the attached output2.log file.

I have been able to build other libraries with the yocto toolchain and cmake with no issues.

Am I missing something?

[output.log](https://github.com/user-attachments/files/17771672/output.log)
[output2.log](https://github.com/user-attachments/files/17771673/output2.log)

**Comment by martxel** [comment, marked as answer]
It seems the issue is that when building and linking aws-iot-device-sdk-cpp-v2, it can't find the s2n library that has been built in the first step.

It may be related to the cmake toolchain file.

As a temporary fix, I have modified the cmake toolchain file and appended the following:
```
list(APPEND CMAKE_FIND_ROOT_PATH "/home/user/src/src/aws-iot-device-sdk-cpp-v2/build/deps/S2N/install/")
```

This way, the find_package works and finds s2n, and the library builds and installs correctly.

**Reply by jmklix** [reply]
It looks like you got it working by adding the correct path to the cmake file. What type of permeant fix are you expecting? Cross compiling can be a very complex process depending on the source and target platforms, so we don't have any specific recommendations if you already have it working for your use case.

**Reply by martxel** [reply]
Honestly, I don't have a clue as to what could be a good fix for this, I don't have a very deep knowledge of cmake and not really sure as to what can be done.
